staging/src/k8s.io/apiserver/plugin/pkg/audit/webhook/webhook.go

```diff
@@ -175,50 +182,105 @@ type batchBackend struct {
 	// Receiving maxBatchSize events will always trigger a send, regardless of
 	// if this amount of time has been reached.
 	maxBatchWait time.Duration
+
+	// Channel to signal that the sending routine has stopped and therefore
+	// it's safe to assume that no new requests will be initiated.
+	shutdownCh chan struct{}
+
+	// The sending routine locks reqMutex for reading before initiating a new
+	// goroutine to send a request. This goroutine then unlocks reqMutex for
+	// reading when completed. The Shutdown method locks reqMutex for writing
+	// after the sending routine has exited. When reqMutex is locked for writing,
+	// all requests has been completed and no new will be spawned, since the
+	// sending routine is not running anymore.
+	reqMutex sync.RWMutex
 }
 
 func (b *batchBackend) Run(stopCh <-chan struct{}) error {
-	f := func() {
-		// Recover from any panics caused by this method so a panic in the
-		// goroutine can't bring down the main routine.
-		defer runtime.HandleCrash()
+	go func() {
+		// Signal that the sending routine has exited.
+		defer close(b.shutdownCh)
 
-		t := time.NewTimer(b.maxBatchWait)
-		defer t.Stop() // Release ticker resources
+		b.runSendingRoutine(stopCh)
 
-		b.sendBatchEvents(stopCh, t.C)
-	}
-
-	go func() {
+		// Handle the events that were received after the last buffer
+		// scraping and before this line. Since the buffer is closed, no new
+		// events will come through.
 		for {
-			f()
-
-			select {
-			case <-stopCh:
-				return
-			default:
+			if last := func() bool {
+				// Recover to try to send all remaining events. Note, that in case of
```

Comment:
Wording: "Attempt to send all remaining events..." That's how I interpreted the intent.